Show warning if Tensor.random_()'s from and to are not in [-(2^digits), 2^digits] bounds for floating-point types

pbelevich opened the pull request:
The documentation states that `random_()` samples "from the discrete uniform distribution". Floating-point types can support _discrete_ _uniform_ distribution only within range [-(2^digits), 2^digits], where `digits = std::numeric_limits<fp_type>::digits`, or

- [-(2^53), 2^53] for double
- [-(2^24), 2^24] for double
- [-(2^11), 2^11] for half
- [-(2^8), 2^8] for bfloat16

The worst scenario is when the floating-point type can not represent numbers between `from` and `to`. E.g. 
```
torch.empty(10, dtype=torch.float).random_(16777217, 16777218)
tensor([16777216., 16777216., 16777216., 16777216., 16777216., 16777216.,
        16777216., 16777216., 16777216., 16777216.])
```
Because 16777217 can not be represented in float

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37537 Show warning if Tensor.random_()'s from and to are not in [-(2^digits), 2^digits] bounds for floating-point types**

Differential Revision: [D21380387](https://our.internmc.facebook.com/intern/diff/D21380387)

pbelevich committed:
Enforce Tensor.random_ check that from and to are in [-(2^digits), 2^digits] bounds for floating-point types

[ghstack-poisoned]

pbelevich committed:
Update on "Enforce Tensor.random_ check that from and to are in [-(2^digits), 2^digits] bounds for floating-point types"




[ghstack-poisoned]

pbelevich committed:
Update on "Enforce Tensor.random_ check that from and to are in [-(2^digits), 2^digits] bounds for floating-point types"




[ghstack-poisoned]

pbelevich committed:
Update on "Enforce Tensor.random_ check that from and to are in [-(2^digits), 2^digits] bounds for floating-point types"

The documentation states that `random_()` samples "from the discrete uniform distribution". Floating-point types can support _discrete_ _uniform_ distribution only within range [-(2^digits), 2^digits], where `digits` = std::numeric_limits<fp_type>::digits`, or

- [-(2^53), 2^53] for double
- [-(2^24), 2^24] for double
- [-(2^11), 2^11] for half
- [-(2^8), 2^8] for bfloat16



[ghstack-poisoned]

pbelevich committed:
Update on "Show warning if Tensor.random_()'s from and to are not in [-(2^digits), 2^digits] bounds for floating-point types"




[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/DistributionTemplates.h:
```diff
@@ -69,6 +69,11 @@ at::Tensor& random_impl(at::Tensor& self, c10::optional<Generator> generator) {
 
+#define CHECK_OUT_OF_BOUNDS_AND_SHOW_WARNING(var, name, min, max, dtype_min, dtype_max) \
+  if (var < min || var > max) { \
+    TORCH_WARN(name , " is out of bounds [", dtype_min, ", ", dtype_max, "]. This warning will become an error in version 1.7 release, please fix the code in advance"); \
```
I think it would be better if you explained to users why this was out of bounds, because your new ranges are technically within the representable range for floating points, and the user will have to know something about floating point numbers to know what went wrong.

ezyang replied:
And maybe a more consistent name for this macro would be `WARN_OUT_OF_BOUNDS`

pbelevich replied:
addressed

ezyang commented on aten/src/ATen/native/DistributionTemplates.h:
```diff
@@ -77,6 +82,16 @@ static void check_from_to_in_range(int64_t from, int64_t to_inc, caffe2::TypeMet
+      const auto two_pow_digits = 1LL << digits;
+      const auto minus_two_pow_digits = -two_pow_digits;
+      const auto two_pow_digits_name = "2^" + std::to_string(digits);
+      const auto minus_two_pow_digits_name = "-(2^" + std::to_string(digits) + ")";
```
Can all of these be made constexpr? If not, it seems like it would be better to ensure that we only render the strings inside of the failing if statement

pbelevich replied:
addressed

pbelevich committed:
Update on "Show warning if Tensor.random_()'s from and to are not in [-(2^digits), 2^digits] bounds for floating-point types"

The documentation states that `random_()` samples "from the discrete uniform distribution". Floating-point types can support _discrete_ _uniform_ distribution only within range [-(2^digits), 2^digits], where `digits = std::numeric_limits<fp_type>::digits`, or

- [-(2^53), 2^53] for double
- [-(2^24), 2^24] for double
- [-(2^11), 2^11] for half
- [-(2^8), 2^8] for bfloat16

The worst scenario is when the floating-point type can not represent numbers between `from` and `to`. E.g. 
```
torch.empty(10, dtype=torch.float).random_(16777217, 16777218)
tensor([16777216., 16777216., 16777216., 16777216., 16777216., 16777216.,
        16777216., 16777216., 16777216., 16777216.])
```
Because 16777217 can not be represented in float




[ghstack-poisoned]

pbelevich committed:
Update on "Show warning if Tensor.random_()'s from and to are not in [-(2^digits), 2^digits] bounds for floating-point types"

The documentation states that `random_()` samples "from the discrete uniform distribution". Floating-point types can support _discrete_ _uniform_ distribution only within range [-(2^digits), 2^digits], where `digits = std::numeric_limits<fp_type>::digits`, or

- [-(2^53), 2^53] for double
- [-(2^24), 2^24] for double
- [-(2^11), 2^11] for half
- [-(2^8), 2^8] for bfloat16

The worst scenario is when the floating-point type can not represent numbers between `from` and `to`. E.g. 
```
torch.empty(10, dtype=torch.float).random_(16777217, 16777218)
tensor([16777216., 16777216., 16777216., 16777216., 16777216., 16777216.,
        16777216., 16777216., 16777216., 16777216.])
```
Because 16777217 can not be represented in float




[ghstack-poisoned]

ezyang commented:
I kind of wonder, in situations like this where you won't get the expected behavior, but some well behaved *does* happen, whether or not it's worth telling people how to work around it if, no actually, they wanted the real behavior. Probably not here though.

pbelevich committed:
Update on "Show warning if Tensor.random_()'s from and to are not in [-(2^digits), 2^digits] bounds for floating-point types"

The documentation states that `random_()` samples "from the discrete uniform distribution". Floating-point types can support _discrete_ _uniform_ distribution only within range [-(2^digits), 2^digits], where `digits = std::numeric_limits<fp_type>::digits`, or

- [-(2^53), 2^53] for double
- [-(2^24), 2^24] for double
- [-(2^11), 2^11] for half
- [-(2^8), 2^8] for bfloat16

The worst scenario is when the floating-point type can not represent numbers between `from` and `to`. E.g. 
```
torch.empty(10, dtype=torch.float).random_(16777217, 16777218)
tensor([16777216., 16777216., 16777216., 16777216., 16777216., 16777216.,
        16777216., 16777216., 16777216., 16777216.])
```
Because 16777217 can not be represented in float


Differential Revision: [D21380387](https://our.internmc.facebook.com/intern/diff/D21380387)

[ghstack-poisoned]
